nixos/weechat: add module

Ma27 opened the pull request:
###### Motivation for this change

Follow-up for #33523. Removed static uid/gid as proposed here (https://github.com/nixos/nixpkgs/pull/33523#issuecomment-372455264) and added support for loading custom scripts.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

lukateras committed:
weechat: add NixOS module

Ma27 commented:
@timokau after yesterday's discussion I was motivated to have a look at it. Seems as it's sufficient to load scripts by specifying their absolute paths (means, store paths :D). Would be cool if you could have a look at this.

@yegortimoshenko as I wanted to hack on a weechat module for NixOS I used your PR from January as base. It would be awesome if you could have a look at this! :)

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=bd891065-4936-4ef1-bd76-eba27eef8930)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/zy4228z6ya2b8k95kfiyb44n96k5r3gq-guile-2.0.13' from 'https://cache.nixos.org'...
copying path '/nix/store/p4pwl0qb7g5h5zg2nqawki8grizl4plf-lua-5.2.3' from 'https://cache.nixos.org'...
copying path '/nix/store/54w9xzm9jrx9xk5bh11033hl4yikkldq-tcl-8.6.6' from 'https://cache.nixos.org'...
copying path '/nix/store/6gmslna2s5ig360raxcpcza289v1bbn0-weechat-2.1-lua' from 'https://cache.nixos.org'...
copying path '/nix/store/rz4zbcdwyd3rg366w48w5g6r24kf0lbd-weechat-2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/wfycvzamj8wyccc4kis0airlhdl71xzs-weechat-2.1-tcl' from 'https://cache.nixos.org'...
copying path '/nix/store/qr8py347wqkmf9r0phq7ixmpflzgrf7n-weechat-2.1-guile' from 'https://cache.nixos.org'...
copying path '/nix/store/nv0006xh0chz5y0fv7ggim5yra1bkqb2-weechat-plugins' from 'https://cache.nixos.org'...
building '/nix/store/gz0158jcb98mxwn251fpvbh5y7h9dwp9-weechat.drv'...
/nix/store/32phnn0rbgfwj8plhzfg0fjhvham4p7w-weechat
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=4b596cb4-ac63-4fd7-993b-2c37f20b0efa)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/xkzj19d9rlapr9d6kf8igw8303qp2wfp-ruby-2.4.4' from 'https://cache.nixos.org'...
copying path '/nix/store/0r8kszrm1kkimdksn85vfbk7cazacpzq-weechat-2.1-guile' from 'https://cache.nixos.org'...
copying path '/nix/store/18p3klp6m1wvf386rxxn81dy9qkvy94q-weechat-2.1-lua' from 'https://cache.nixos.org'...
copying path '/nix/store/jp7r4n062yh20v6gr6m924l57n6grjcm-weechat-2.1-perl' from 'https://cache.nixos.org'...
copying path '/nix/store/bac623nwawy37l1yndq8iqns7a98gffj-weechat-2.1-python' from 'https://cache.nixos.org'...
copying path '/nix/store/silpcbg8k5bghgilrdvdvw21kvfzzmbz-weechat-2.1-ruby' from 'https://cache.nixos.org'...
copying path '/nix/store/hkx1zhx72xxm33313lb7ars4f58ghkrj-weechat-2.1-tcl' from 'https://cache.nixos.org'...
copying path '/nix/store/bvxka67lyagz9igy22m8mi3mpj5n3rq9-weechat-plugins' from 'https://cache.nixos.org'...
building '/nix/store/vhcbc7za2lbkmclwvrdgprl8qdarcg4w-weechat.drv'...
/nix/store/iq75mvwy26qf28lhax19867h4i951nym-weechat
```
</p></details>

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+        options = {
+          name = mkOption {
+            type = str;
+            description = "Path to the script in the derivation.";
```
So `name` is actually a path?

Ma27 replied:
I was thinking about "name of the script", but well, there are only two hard things in computer science, right? ;)

But seriously, you're right, `path` is much better, however this can be dropped as you stated later on.

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+      default = [];
+      example = literalExample ''
+        [
+          { name = "share/jabber.py"; script = pkgs.weechat-xmpp }
```
Why not just `"${pkgs.weechat-xmpp}/share/jabber.py"`? Or maybe even force the script packages to be in a standard format and only require the package.

Ma27 replied:
you're absolutely right! I originally wanted to define a pre-defined output structure for derivations which feature weechat scripts, but dropped the idea later and kept the (admittedly too complex submodule type).

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+      '';
+    };
+    sessionName = mkOption {
+      description = "Name of the `screen' session for weechat.";
```
I think it is better to use normal quotes here. I don't think the tex-like backtick/normal tick combination is very common in plain text.

Ma27 replied:
not sure though. I've seen this fairly often in the Nix(OS) context including descriptions for options and wanted to be convenient. But feel free to correct me, I'm not sure about this :)

timokau replied:
You're right, I can find a couple of examples of both simple quotes and `' quotes in `man configuration.nix`. I personally prefer simple quotes but since there is apparently no consensus I'm fine with whatever you prefer.

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+      wantedBy = [ "multi-user.target" ];
+    };
+
+    services.weechat.init = mkIf (builtins.length cfg.scripts != 0) (concatStringsSep "\n"
```
Is this mkif really necessary? Wouldn't it just be an empty string in the case of length 0?

Ma27 replied:
You're absolutely right, this makes the expression way easier to read, thanks!

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+    };
+
+    services.weechat.init = mkIf (builtins.length cfg.scripts != 0) (concatStringsSep "\n"
+      (map (script: "/script load ${script.script}/${script.name}\n") cfg.scripts));
```
I don't have very much experience with the module system and merging and everything -- will these scripts be loaded *before* the rest of the init code is executed? I think they should be.

Ma27 replied:
I'm not sure either, but you're right, it to avoid breaking scripts it might be helpful to use `mkBefore` to be absolutely sure :)

timokau commented on nixos/modules/services/misc/weechat.nix:
```diff
@@ -0,0 +1,103 @@
+    systemd.services.weechat-apply-init = let
+      initFile = pkgs.writeText "weechat-init" cfg.init;
+    in {
+      script = "sed 's/^/*/' ${initFile} > ${cfg.root}/weechat_fifo";
```
What does this do?

Ma27 replied:
according to the docs (https://weechat.org/files/doc/stable/weechat_user.en.html#fifo_plugin), the `*` at the beginning of a string is needed when passing commands to weechat through `weechat_fifo`. This was actually introduced in the previous commit I didn't author, but it's more convenient to type `/cmd args...` without the star at the beginning.

Ma27 replied:
I just used the code from the previous commit. However I think that this is the right way to go as the star is only used with `fifo`.
However the `/` at the beginning is also used when running weechat from an active session, so it's IMHO more convenient to keep this.

timokau commented:
Thanks! Any reason you went with a nixos module? I feel like adding the configuration through a package wrapper may be nicer.

Ma27 commented:
> Any reason you went with a nixos module? I feel like adding the configuration through a package wrapper may be nicer.

I originally wanted to make the weechat module of @yegortimoshenko ready to merge and wanted to see if it's possible to easily add support for scripts (which is easily doable by adding `/script load $path` to `services.weechat.init`). Previously I assumed that scripts are only loadable if they're installed in `~/.weechat`, however absolute paths are possible as well.

I'm not sure if we want to duplicate the `weechat_fifo` the logic of the previously written module in the package (and if it's even possible), but I'm happy to discuss about it to see which is the best decision for nixpkgs :)

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=36eed414-76f4-4d2a-bfd2-37a051cfe6c0)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/p4pwl0qb7g5h5zg2nqawki8grizl4plf-lua-5.2.3' from 'https://cache.nixos.org'...
copying path '/nix/store/54w9xzm9jrx9xk5bh11033hl4yikkldq-tcl-8.6.6' from 'https://cache.nixos.org'...
copying path '/nix/store/6gmslna2s5ig360raxcpcza289v1bbn0-weechat-2.1-lua' from 'https://cache.nixos.org'...
copying path '/nix/store/rz4zbcdwyd3rg366w48w5g6r24kf0lbd-weechat-2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/wfycvzamj8wyccc4kis0airlhdl71xzs-weechat-2.1-tcl' from 'https://cache.nixos.org'...
copying path '/nix/store/zy4228z6ya2b8k95kfiyb44n96k5r3gq-guile-2.0.13' from 'https://cache.nixos.org'...
copying path '/nix/store/qr8py347wqkmf9r0phq7ixmpflzgrf7n-weechat-2.1-guile' from 'https://cache.nixos.org'...
copying path '/nix/store/nv0006xh0chz5y0fv7ggim5yra1bkqb2-weechat-plugins' from 'https://cache.nixos.org'...
building '/nix/store/gz0158jcb98mxwn251fpvbh5y7h9dwp9-weechat.drv'...
/nix/store/32phnn0rbgfwj8plhzfg0fjhvham4p7w-weechat
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=8d4f45fc-e612-4d86-af53-1ea239dbd08d)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/iq75mvwy26qf28lhax19867h4i951nym-weechat
```
</p></details>

Ma27 commented:
I learned about `weechat_fifo` as an approach to "remotely control" weechat. You could even run a systemd unit during the runtime of `weechat` with new options.

However `--run-command` seems to be a reasonable alternative. Please give me some time to think about this, then we can continue the discussion:)

Ma27 committed:
weechat: 2.1 -> 2.2; improve package configuration

This aims to make the `weechat` package even more configurable. It
allows to specify scripts and commands using the `configure` function
inside a `weechat.override` expression.

The package can be configured like this:

```
with import <nixpkgs> { };
weechat.override {
  plugins = { availablePlugins, ... }: {
    plugins = builtins.attrValues availablePlugins;

    init = ''
      /set foo bar
      /server add freenode chat.freenode.org
    '';

    scripts = [ "/path/to/script.py" ];
  };
}
```

All commands are passed to `weechat --run-command "/set foo bar;/server ..."`.

The `plugins' attribute is not necessarily required anymore, if it's
sufficient to add `init' commands, the `plugins' will be
`builtins.attrValues availablePlugins' by default.

Additionally the result contains `weechat` and `weechat-headless`
(introduced in WeeChat 2.1) now.

Ma27 committed:
nixos/weechat: cleanup module, add module documentation

This adds several improvements the previously introduced
`services.weechat` module:

* Dropped `services.weechat.init` as the initialization script can now
  be done on package-level since 2af41719bc using the `configure`
  function.

* Added `sessionName` option to explicitly configure a name for the
  `screen` session (by default: weechat-screen).

* Added `binary` option to configure the binary name (e.g.
  `weechat-headless`).

* Added docs regarding `screen` session and `weechat.service`.

Ma27 commented:
btw, do you think we can sneak this into 18.09? :)

timokau commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -117,13 +118,25 @@ in if configure == null then weechat else
+      #!${stdenv.shell}
+      export WEECHAT_EXTRA_LIBDIR=${pluginsDir}
+      ${lib.concatMapStringsSep "\n" (p: lib.optionalString (p ? extraEnv) p.extraEnv) plugins}
+      exec ${weechat}/bin/${bin} --run-command "${init}" "$@"
```
If subsequent `--run-command`s overwrite the first one, maybe we could check for `--run-command` or `-r` in `$@` and concat the two. It would add a bit of complexity but just silently overwriting may confuse users.

timokau reviewed (changes requested):
Awesome, thanks! There is always a bit to nitpick, but looks good overall.

timokau commented:
> 18.09

Not sure what the policy for the "feature freeze" is. I'd be in favor (since this doesn't update weechat and is unlikely to break anything) but we should ask one of the release managers once this is merged.

Ma27 commented on doc/package-notes.xml:
```diff
@@ -704,6 +704,42 @@ overrides = super: self: rec {
+  };
+}</programlisting>
+    These values can be overridden manually when running
+    <literal>weechat --run-command "your-commands"</literal>.
```
unfortunately not. I tried it out with multiple `--run-command` arguments. I can try to merge both values though, but I'm not sure, how easy this will be...

timokau replied:
It *is* a bit of complexity (especially for a bash script), but I think it is worth it because the alternative can be very confusing. This should do the trick (assuming `init` is set correctly):

```bash
#!/bin/sh
init="" # init script

new_args=()
# remove --run-command arguments and remember their commands
while [ $# -gt 0 ]; do
	if [[ "$1" == "--run-command" ]] || [[ "$1" == "-r" ]]; then
		# subsequent `--run-commands` override the previous ones, as is default weechat behaviour
		commands_to_run="$2"
		shift
	else
		new_args+=("$1")
	fi
	shift
done

new_args+=("--run-command" "$init;$commands_to_run")

exec weechat "${new_args[@]}"

```

timokau replied:
Oh I just saw that you already implemented that.

Ma27 commented on doc/package-notes.xml:
```diff
@@ -704,6 +704,42 @@ overrides = super: self: rec {
+  </para>
+  <para>
+    The addition of the <literal>init</literal> parameter to the <literal>override</literal> API
+    of WeeChat allows <literal>Nix</literal> to generate packages that autoconfigure themselves.
```
You can set values for your scripts before weechat actually starts. But you're right, re-reading it makes the  sentence quite odd :sweat_smile:

timokau replied:
So basically just "you can configure WeeChat"? I think the formulation should be improved, at least I'm having a hard time with it.

Ma27 commented:
@timokau thanks a lot for your feedback! Would you mind having a look at this a second time? :)

timokau commented on doc/package-notes.xml:
```diff
@@ -704,6 +704,42 @@ overrides = super: self: rec {
+    The <literal>configure</literal> method can be used to pass commands to the program:
+<programlisting>weechat.override {
+  configure = { availablePlugins, ... }: {
+    plugins = builtins.attrValues availablePlugins;
```
Oh right, I forgot that plugins and scripts are two entirely different things. Maybe that should be made explicit in the docs. So this actually just adds all available plugins?

Ma27 replied:
You're right, the terms `script` and `plugin` seem to be the same thing, I made the mistake several times as well.

However I don't think that it's needed to add another sentence here, when you have a look at the [WeeChat docs in the `nixpkgs` manual](https://github.com/NixOS/nixpkgs/blob/2eea92a1dc124e24744d57f53d9dbe10e85876e1/doc/package-notes.xml#L662), you'll see that how to configure and use `plugins` with the Nix package is explained in the previous paragraph.

timokau replied:
Right, its easy to get lost in diffs and miss the context. However I think a short aside like "note that plugins are not the same as scripts. See the scripts section for more information" or something when plugins are first introduced may be useful.

timokau replied:
Also does that mean that no plugins are installed if I pass in a config without specifying plugins? The default (all plugins) should probably be kept if plugins aren't specified.

timokau replied:
Why not change `inherit (config) plugins` to `plugins = config.plugins or availablePlugins`? Requiring that argument made sense when it was the only argument but makes less sense now. I suspect most people don't care much about a couple of MB in the closure size of an end user application.

timokau commented on doc/package-notes.xml:
```diff
@@ -704,6 +704,42 @@ overrides = super: self: rec {
 </programlisting>
   </para>
+  <para>
+    WeeChat allows to set defaults on startup using the <literal>--run-command</literal>.
```
Alright

timokau commented on doc/package-notes.xml:
```diff
@@ -704,6 +704,42 @@ overrides = super: self: rec {
+    plugins = builtins.attrValues availablePlugins;
+
+    scripts = [
+      "${pkgs.weechat-xmpp}/share/jabber.py"
```
Yeah something like

```
scripts  = with pkgs.weechatScripts [
  xmpp
  matrix-protocol
]
```

Would be pretty neat. It would be possible to remain backwards compatibility, although I'm not sure if its worth it.

Ma27 replied:
I'm not sure, but I'd even say that this is out of scope. This PR already includes script/config support for the package *and* a module and I'd say that splitting single scripts into their own metapackage should be done in a second PR.

Regarding the change itself: I'm definetely in favor of it, users don't need to know about the structure inside a given store path, they simply need to know their packages, furthermore this is way more readable IMHO.

Regarding backwards compatibility: I've seen several people using these packages, so I think that just moving them is a bad idea. Let's place them into `pkgs/top-level/aliases.nix` and print a deprecation warning if this attribute path is used.

timokau replied:
Scripts outside of the nix store would be an impurity that would make the config less portable. I see how it could be useful however, so if you want to support that its fine by me.

Ma27 replied:
I guess you're right, packaging weechat scripts shouldn't be too hard.I'll see how much time it costs to implement this, then I may add this to this PR.

timokau commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -117,13 +118,34 @@ in if configure == null then weechat else
+      for i in "$@"; do
+        ((index+=1))
+        if [[ "$i" = "--run-command" || "$i" = "-r" ]]; then
+          next="$next;''${list[$index]}"
```
We probably shouldn't modify weechats standard behaviour of overriding previous `--run-commands` (that were explicitly passed by the user). Although I personally think aggregating them would make way more sense.

On the other hand weechat may change that in the future and there isn't really too much harm in aggregating them (who in their right mind would depend on the override behaviour?), so I'm not sure.

timokau replied:
I've opened https://github.com/weechat/weechat/issues/1248 upstream for that.

timokau replied:
You currently append to `next` each time you find a `--run-command`, which is different from the default weechat behaviour.

However somebody upstream already said they'll "implement that quickly", so it should be fine as-is.

Ma27 committed:
weechatScripts.wee-slack: init at 2.1.1

Ma27 commented:
@timokau I added myself as maintainer and introduced the `weechatScripts` metapackage to load scripts on startup. Would you mind having *another* look? :)

timokau commented on doc/package-notes.xml:
```diff
@@ -704,6 +706,57 @@ overrides = super: self: rec {
+  };
+}</programlisting>
+    Please keep in mind that <literal>scripts</literal> doesn't require store paths.
+    WeeChat simply needs to be able to find the specified scripts.
```
This part is out of date now right?

timokau commented on doc/package-notes.xml:
```diff
@@ -704,6 +706,57 @@ overrides = super: self: rec {
+  };
+  passthru.scripts = [ "foo.py" "bar.lua" ];
+  installPhase = ''
+    mkdir $out/share
```
I would prefer those variables quoted, although there was some contoversy about that. So I won't insist on it.
Also applies to other files changed here.

Ma27 replied:
Admittedly I've never seen this pattern. Furthermore it's IIRC documented in Nixpkgs (or even Nix?), that `$out` exists as reference to the pending store path.

As you mention "some controversy about that", would you mind sending me the link to the discussion? I don't think that this is needed, but I'd like to see the arguments of both sides :)

timokau replied:
Yes, `$out` is a reference to the pending store path. I think as a shell variable it should still be quoted. I was referring to this discussion: https://github.com/NixOS/nixpkgs/pull/45657#discussion_r212835434

timokau commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -78,52 +79,77 @@ let
+      index=0
+      next=
+      declare -a list=("$@")
+      for i in "$@"; do
```
A short comment that explains what this does would be nice.

timokau commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -78,52 +79,77 @@ let
+      done
+      export WEECHAT_EXTRA_LIBDIR=${pluginsDir}
+      ${lib.concatMapStringsSep "\n" (p: lib.optionalString (p ? extraEnv) p.extraEnv) plugins}
+      exec ${weechat}/bin/${bin} "$@" --run-command "$next;${init}"
```
This would double the `--run-commands` if https://github.com/weechat/weechat/issues/1248#issuecomment-419056035 is implemented right? If so, you should add a comment to the `version` attribute reminding whoever updates weechat to adjust that.

timokau reviewed (commented):
Looks good, nearly done :)

timokau commented on pkgs/applications/networking/irc/weechat/default.nix:
```diff
@@ -78,52 +79,83 @@ let
+      declare -a list=("$@")
+
+      # check if $@ contains a `--run-command` option. If so,
+      # it will be merged with the `init' property from `weechat.override'.
```
Now that this is [fixed upstream](https://github.com/weechat/weechat/issues/1248#event-1831302139), we could instead cherry-pick the upstream solution as a patch. Either that or add a comment to the `version` attribute to remind whoever updates weechat next to remove this.

Ma27 replied:
now this was fast :D 
Unfortunately the patch doesn't apply onto v2.2, I'll try to rebase it, but it seems to be a bit of work 
as several translation files are affected.

timokau replied:
I was definitely positively surprised, nice when upstream is so reactive :)

Yeah no reason to rebase it I think. You could make use of fetchpatch's `includes` if the translation files are the only issues (that really are a *lot* of translation files). Otherwise the comment to the version is also fine.

Ma27 replied:
ah dammit, `src/core/weechat.c` contains some conflicts as well. But I'll think about this tomorrow, I'll catch some sleep now :)

timokau replied:
No need to spend too much time on this then. We just need to make sure that the code doesn't stay around forever and for that the comment should be enough.

timokau commented:
Thanks! Unfortunately ofBorg is reporting

```
undefined variable 'weechat' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/nixos/modules/services/misc/weechat.nix:23:56
```

GrahamcOfBorg commented:
<!--REQUEST_ID=b2e52ad6-903b-4849-9c69-a35e6b1bf129-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=848a839a-7752-4014-ab3f-cd254ba18f44)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/y4ymnvgxygpq05h03kyzbj572zmh6zla-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/1c3drc6cfx3w0xa4hai4ii3gnyb8fhf2-weechat-2.2-python/lib
patching script interpreter paths in /nix/store/1c3drc6cfx3w0xa4hai4ii3gnyb8fhf2-weechat-2.2-python
checking for references to /build in /nix/store/1c3drc6cfx3w0xa4hai4ii3gnyb8fhf2-weechat-2.2-python...
building '/nix/store/i3afih08s06c5wb9b9jsj00nck3wyaxg-weechat-plugins.drv'...
building '/nix/store/bj3ibijv09dh8q224n0xc268dln0xyib-weechat-headless.drv'...
building '/nix/store/5gay07v7b8qhdjiv4if2hshq03fj8igg-weechat.drv'...
building '/nix/store/2bcyrzs2k9rsg074qnmvcsq0f3d92r7r-weechat-bin-env.drv'...
created 2 symlinks in user environment
/nix/store/gkmkgwr85fsyxw147c90rvajqij2pl7r-weechat-bin-env
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=b2e52ad6-903b-4849-9c69-a35e6b1bf129-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45728&attempt_id=be31a0c1-d2ee-4777-8bf1-e2cd5b0b175c)

Attempted: weechat

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/h0lbngpv6ln56hjj59i6l77vxq25flbz-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/ncgf7b9c9iiflw8qngi2argvl749zr13-weechat-2.2-python/lib
patching script interpreter paths in /nix/store/ncgf7b9c9iiflw8qngi2argvl749zr13-weechat-2.2-python
checking for references to /build in /nix/store/ncgf7b9c9iiflw8qngi2argvl749zr13-weechat-2.2-python...
building '/nix/store/02y5rfbm4q4smv92qa3xb31a65zv165v-weechat-plugins.drv'...
building '/nix/store/2wnb41yg1s6bbwaz6i13wkxm304awya8-weechat-headless.drv'...
building '/nix/store/f2kfwhh1f2hj519w2likc6nsj9imr9lr-weechat.drv'...
building '/nix/store/6pg1xpiskkslqw1g9jbxfdky7m9rcdvs-weechat-bin-env.drv'...
created 2 symlinks in user environment
/nix/store/5sx11mjwaq4mil6mm25k36kwg949blrs-weechat-bin-env
```
</p></details>

timokau commented:
Great, thank you for your patience :) Looks good now.

Just FYI, GitHub doesn't send email notifications when you force-push. I only noticed because of ofBorgs rebuilds.

timokau commented:
#46324 should also be included in 18.09 if we include this.

samueldr commented:
Oof, the backport policies are really about updating software, and updating software revisions for non-major updates. I don't know if a new module should be backported. it's kind of a lot of things, and the package has deviated considerably compared to how it worked prior to this PR.

SInce nix, nixos and modules can be composed pretty easily, I'm not sure this should be backported. Though the *version update* (only) could and probably should be backported. (Though it does not seem to have security fixes, it does have stability fixes.)

I'm not strictly against backporting this, but I'm not sure it should be. I'm don'T know about previous precedents of backports of non-trivial module updates. And then this would create such a precedent.

Thoughts, @vcunat?

timokau commented:
On 18-09-07 19:47, Samuel Dionne-Riel wrote:
> Oof, the backport policies are really about updating software, and updating software revisions for non-major updates. I don't know if a new module should be backported. it's kind of a lot of things, and the package has deviated considerably compared to how it worked prior to this PR.

Yes but the most "risky" changes are all additions, so no regressions. And since weechat is an end-user application and not a library, potential for breakage is limited and can be tested pretty quickly.

> SInce nix, nixos and modules can be composed pretty easily, I'm not sure this should be backported.

Can they? I haven't found a good way to compose modules from different NixOS versions yet. For me the most important part of this is the scripts support, which can be used by simply fetching the weechat package from unstable. So I don't care very strongly about backporting it, but I still think it would be better. I understand the arguments against it however -- feature freeze is feature freeze.

> -- 
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub:
> https://github.com/NixOS/nixpkgs/pull/45728#issuecomment-419607203

timokau commented:
Note that while #46324 was a breakage, it was not a regression. Just a small bug in the newly added features.

vcunat commented:
@samueldr: it's just a new module, right?  Those who don't enable it shouldn't be affected at all by that.

timokau commented:
A new module and configurability added to the weechat package. Still shouldn't affect anybody not using it.

samueldr commented:
Sorry, I meant to reply earlier.

I must have misread that it was changing an existing module. Probably by relying on the description "removing [...] adding support for [...]" instead of looking at the actual code.

In that case it is a fine candidate for a backport I guess.

Ma27 commented:
> Sorry, I meant to reply earlier.

No worries, we've learned at last NixCon that release management is quite a hard and time-consuming task :)

> In that case it is a fine candidate for a backport I guess.

great! I'm currently using the changes with a recent 18.09 checkout personally, I'll recheck and file a PR later to do the backport, ok?

samueldr commented:
Yes, you can tag me in the backport PR. Thanks for making the PR!

dtzWill commented:
Hey--how is this intended to be used? Anyone have an example?

The documentation suggests this is geared so that you can attach to the screen session (command to do so is mentioned, and modification to screenrc suggested, etc.), but that is a non-starter AFAICT since screen needs to be suid for that to work and our screen is not.

Which makes me wonder--given this was merged so recently... how are folks using it? Or is there a way to attach I'm missing (sorry!)? Or to run as a user service, perhaps? Thanks!

Ma27 commented:
Hey! Thanks a lot for you interest in this module!
Unfortunately I'll a bit busy in the next days, so please give me some time, then I'll have a deeper look at the issue, ok? :)

dtzWill commented:
Sure thing! No hurry at all :).  See you in a few!

Ma27 commented:
first of all sorry for the delay, I meant write earlier.

After debugging I realised that I used a `screen` with `setuid` on my server (as I needed it for some experiments earlier). During next week I'll whats the best way to implement this and file a patch:)
